src/dense.jl

```diff
@@ -1610,6 +1580,66 @@ factorize(A::Adjoint)   =   adjoint(factorize(parent(A)))
 factorize(A::Transpose) = transpose(factorize(parent(A)))
 factorize(a::Number)    = a # same as how factorize behaves on Diagonal types
 
+function getstructure(A::StridedMatrix)
+    m, n = size(A)
+    if m == 1 return A[1] end
+    utri    = true
+    utri1   = true
+    herm    = true
+    sym     = true
+    for j = 1:n-1, i = j+1:m
+        if utri1
+            if A[i,j] != 0
+                utri1 = i == j + 1
+                utri = false
+            end
+        end
+        if sym
+            sym &= A[i,j] == A[j,i]
+        end
+        if herm
+            herm &= A[i,j] == conj(A[j,i])
+        end
+        if !(utri1|herm|sym) break end
+    end
+    ltri = true
+    ltri1 = true
+    for j = 3:n, i = 1:j-2
+        ltri1 &= A[i,j] == 0
+        if !ltri1 break end
+    end
+    if ltri1
+        for i = 1:n-1
+            if A[i,i+1] != 0
+                ltri &= false
+                break
+            end
+        end
+    end
+    return (utri, utri1, ltri, ltri1, sym, herm)
+end
+_check_sym_herm(A) = (issymmetric(A), ishermitian(A))
+_check_sym_herm(A::AbstractMatrix{<:Real}) = (sym = issymmetric(A); (sym,sym))
+function getstructure(A::AbstractMatrix)
+    utri1 = istriu(A,-1)
+    # utri = istriu(A), but since we've already checked istriu(A,-1),
+    # we only need to check that the subdiagonal band is zero
+    utri = utri1 && iszero(diag(A,-1))
```

Comment:
Is there any value in trying to avoid the allocation of the subdiagonal, with a view or `diagind` or something?

Reply:
I'd say that in this case it's perhaps better to allocate, as the `diag` may be a special array type (e.g. a sparse array), whereas `iszero` for the view may dispatch to a generic fallback method.